### ISSUE TITLE
Suggest a more friendly way to generate

### ISSUE THREAD

**Issue by oovm** [Cysharp/csbindgen#22]
I want to export a complex structure, but its internal definition does not need to be exposed to Csharp, and all its' interface using csharp type.

https://github.com/Cysharp/csbindgen#pointer is what I need, but I want to have some better output.

The flow I want is roughly as follows:

## Step1, Rust High Level

```rust
#[repr(C)]
pub struct CounterContext {
    set: HashSet<i32>,
}

#[cs_bindgen(Box)]  // or Rc, Arc, ...
// replace all self as Box<Self>, then can wrap by Box::into_raw and Box::from_raw
impl CounterContext {
    #[cs_bindgen(constructor)]
    pub fn create_counter_context() -> Self {
        Self {
            set: HashSet::new(),
        }
    }
    // value is a csharp type
    pub fn insert_counter_context(&mut self, value: i32) {
        self.set.insert(value);
    }
    #[cs_bindgen(destructor))]
    pub fn delete_counter_context(&mut self) {
        for value in self.set.iter() {
            println!("counter value: {}", value)
        }
    }
}
```

## Step2, Rust Low Level

```rust
#[no_mangle]
pub extern "C" fn create_counter_context() -> *mut c_void {
    let ctx = Box::new(CounterContext {
        set: HashSet::new(),
    });
    Box::into_raw(ctx) as *mut c_void
}

#[no_mangle]
pub unsafe extern "C" fn insert_counter_context(context: *mut c_void, value: i32) {
    let mut counter = Box::from_raw(context as *mut CounterContext);
    counter.set.insert(value);
    Box::into_raw(counter);
}

#[no_mangle]
pub unsafe extern "C" fn delete_counter_context(context: *mut c_void) {
    let counter = Box::from_raw(context as *mut CounterContext);
    for value in counter.set.iter() {
        println!("counter value: {}", value)
    }
}
```

## Step3, C# Low Level

```cs
    internal static unsafe partial class NativeMethods
    {
        const string __DllName = "nativelib";

        [DllImport(__DllName, EntryPoint = "create_counter_context", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
        public static extern void* create_counter_context();

        [DllImport(__DllName, EntryPoint = "insert_counter_context", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
        public static extern void insert_counter_context(void* context, int value);

        [DllImport(__DllName, EntryPoint = "delete_counter_context", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
        public static extern void delete_counter_context(void* context);

    }
```

## Step4, C# High Level

```cs
    public class CounterContext
    {
        private void* _handle;
        public CounterContext()
        {
            _handle = NativeMethods.create_counter_context();
        }

        public void Insert(int value)
        {
            NativeMethods.insert_counter_context(_handle, value);
        }

        ~CounterContext()
        {
            NativeMethods.delete_counter_context(_handle);
        }
    }
```

This makes a more human-friendly experience.

**Comment by neuecc** [comment]
I think you want something like [SafeHandle](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle) auto-generated, 
but you want the raw Rust support, I think it can only be generated from a specific grammar.
I don't think it would be very practical to have people write Rust while making them aware of such constraints.

At least as far as Rust Low Level is concerned, I believe it is better to have human handwritten control.

As for C# High Level, I have an idea, for example, to generate it by grouping with C# Source Generator.
Generating C++ classes with bindgen will generate a lot of `handler* @this` methods, which will be grouping identifiers.
In other words, I think automatic generation of C# Low Level to C# High Level is a good idea.

**Comment by neuecc** [comment]
I've added SourceGenerator in 1.7.0.
You can use like this.

![image](https://user-images.githubusercontent.com/46207/232473592-7457d946-310a-4053-893b-34b85fd847a8.png)

![image](https://user-images.githubusercontent.com/46207/232474592-7ff3e898-b212-4ed7-a02d-d105396d47e4.png)

**Comment by oovm** [comment]
@neuecc Other features can be used, but I can't find `Unsafe` in my unity environment, even with 

`using System.Runtime.CompilerServices;`

Is it possible to add an option to generate `Unity.Collections.LowLevel.Unsafe.UnsafeUtility`?

**Comment by neuecc** [comment]
Adding CompilerServices.Unsafe.dll will work in Unity. Is there any reason you can't add it?
